### ISSUE TITLE
Native debian package for appimaged

### ISSUE THREAD

**Issue by probonopd** [probonopd/AppImageKit#339]
@rickysarraf  writes in https://github.com/probonopd/AppImageKit/issues/325

> I can speak on what is the case in Debian. I'm not sure about other distributions. ON Debian, on the buildd infrastructure, network access is disabled. And there are good reasons to do that. Your git submodules are just defined as a git repository.

I am not sure what is the best path of action here, but I will definitely support anyone who wants to make a debian package for `appimaged`. Also see the great work by @marguerite who has packaged it (and `appimagetool`) for openSUSE.

Perhaps @ScarlettGatelyClark wants to chime in?

**Comment by marguerite** [comment]
In openSUSE Build Service, we don't have network access in the build vitual machine created dynamically either. But we have something called source service that can git with sub modules and compress the result to a tarbal. So running git submodule directly when building is also impossible here. It has to be commented out for native packaging.

But in bash there are ways to detect network availability. So I think we could use git only when network is possible. on those isolated machines, just unpack the tarball. Leaving the responsibility to create the tarball to the distro packagers

**Comment by marguerite** [comment]
I think checking the common local IP could be enough. ping or nc is too heavy in this case.

**Comment by rickysarraf** [comment]
There are 2 parts of the problem here.

1. Network access. This can be easily solved. Either upstream does point releases, with tarballs, or else I can myself take git snapshots and create a tarball.

2. forked sub-modules: There's a forked squashfs-tools git submodule. Under debian squashfs-tools is already packaged. And I'm not going to carry this forked module inside appimagekit. You could either push your change to the squashfs folks and add a versioned build dependency here. Or you could just carry the delta and ask the user to fetch the squashfs-tools source themselves. I'd suggest you push your changes, because otherwise even if you carry the delta, it becomes difficult for me to push into a pristine source package in Debian.

So to summarize point 2, the concern is not about a sub-module, but more about carrying embedded copies of the source.

The current build script structure is trying too hard being simple for the end user. In doing that, it is breaking standard practices. In my opinion, you should keep them separately as end-user-helper-scripts and instead have a much standardized build setup.

I doubt if end-users are going to build the tool. Most they care is is a binary package.

**Comment by probonopd** [comment]
@rickysarraf 

> There's a forked squashfs-tools git submodule. Under debian squashfs-tools is already packaged. And I'm not going to carry this forked module inside appimagekit

Yes, I am not happy about that.

> You could either push your change to the squashfs folks

Sent a PR half a year ago but no reaction: https://github.com/plougher/squashfs-tools/pull/13 - is something wrong with it?

> Or you could just carry the delta and ask the user to fetch the squashfs-tools source themselves

Do you mean "carry the delta" as in "provide a diff"?

**Comment by rickysarraf** [comment]
On Sun, 2017-01-29 at 03:33 -0800, probonopd wrote:
> Yes, I am not happy about that.
> You could either push your change to the squashfs folks
> Sent a PR half a year ago but no reaction: plougher/squashfs-tools#13 - is
> something wrong with it?

That repo seems inactive. So does the git repo at kernel.org. but I would rely
on that one.

Let's try fresh now. Given there's been not much development in the upstream
repo, I've just taken your patch, as is, for submission.

> Or you could just carry the delta and ask the user to fetch the squashfs-tools 
> source themselves
> Do you mean "carry the delta" as in "provide a diff"?

Doing now.

Dear Squashfs Team,

Can you please review/include the attached patch into squashfs-tools ?
The mentioned changes are needed by AppImageKit.

What is AppImage

The AppImage format is a format for packaging applications in a way that allows
them to run on a variety of different target systems (base operating systems,
distributions) without further modification.

https://en.wikipedia.org/wiki/AppImage

AppImageKit is a concrete implementation of the AppImage format and provides
tools such as appimagetool and appimaged for conveniently handling AppImages.

https://github.com/probonopd/AppImageKit



-- 
Ritesh Raj Sarraf
RESEARCHUT - http://www.researchut.com
"Necessity is the mother of invention."

**Comment by probonopd** [comment]
Thanks @rickysarraf

**Comment by rickysarraf** [comment]
You're welcome. Let's hope it gets merged.

Once that is done, I can care for its (and squashfuse) inclusion into Debian.

**Comment by rickysarraf** [comment]
On Sun, Jan 29, 2017 at 2:00 PM, Ritesh Raj Sarraf <rrs@researchut.com>
wrote:

> On Sun, 2017-01-29 at 03:33 -0800, probonopd wrote:
> > Yes, I am not happy about that.
> > You could either push your change to the squashfs folks
> > Sent a PR half a year ago but no reaction: plougher/squashfs-tools#13 -
> is
> > something wrong with it?
>
> That repo seems inactive. So does the git repo at kernel.org. but I would
> rely
> on that one.
>

The repo at github is dead, I only used it during the time kernel.org was
offline due to the hack.

The repo at kernel.org is the official repo.  It hasn't been updated in a
while simply because the tools are considered stable, and I have not had a
burning desire to add new features for a while.

>
> Let's try fresh now. Given there's been not much development in the
> upstream
> repo, I've just taken your patch, as is, for submission.
>
> > Or you could just carry the delta and ask the user to fetch the
> squashfs-tools
> > source themselves
> > Do you mean "carry the delta" as in "provide a diff"?
>
> Doing now.
>
> Dear Squashfs Team,
>

Well that's me.

>
> Can you please review/include the attached patch into squashfs-tools ?
> The mentioned changes are needed by AppImageKit.
>
>
Well from a quick look at the attached patches, I see no reason not to pull
them.

I'll have a look over at github because is the first I've seen of these
patches, and I'll see if I can pull them from there.

If not, I'll ask the author to send them to me at this address.
Phillip@lougher.demon.co.uk is dead because my ISP arbitrarily ceased
email support sometime ago (and I'd had that email for 20 years).  It would
appear phillip@squashfs,org.uk is still redirected to that, and I'll fix
that.

Phillip



> What is AppImage
>
> The AppImage format is a format for packaging applications in a way that
> allows
> them to run on a variety of different target systems (base operating
> systems,
> distributions) without further modification.
>
> https://en.wikipedia.org/wiki/AppImage
>
> AppImageKit is a concrete implementation of the AppImage format and
> provides
> tools such as appimagetool and appimaged for conveniently handling
> AppImages.
>
> https://github.com/probonopd/AppImageKit
>
>
>
> --
> Ritesh Raj Sarraf
> RESEARCHUT - http://www.researchut.com
> "Necessity is the mother of invention."

**Comment by rickysarraf** [comment]
On Sun, Jan 29, 2017 at 2:00 PM, Ritesh Raj Sarraf <rrs@researchut.com> wrote:
> On Sun, 2017-01-29 at 03:33 -0800, probonopd wrote:
>> Yes, I am not happy about that.
>> You could either push your change to the squashfs folks
>> Sent a PR half a year ago but no reaction: plougher/squashfs-tools#13 - is
>> something wrong with it?
>
> That repo seems inactive. So does the git repo at kernel.org. but I would rely
> on that one.
>

The repo at github is dead, I only used it during the time kernel.org
was offline due to the hack.

The repo at kernel.org is the official repo.  It hasn't been updated in a while
simply because the tools are considered stable, and I have not had a burning
desire to add new features for a while.

> Let's try fresh now. Given there's been not much development in the upstream
> repo, I've just taken your patch, as is, for submission.
>
>> Or you could just carry the delta and ask the user to fetch the squashfs-tools
>> source themselves
>> Do you mean "carry the delta" as in "provide a diff"?
>
> Doing now.
>
> Dear Squashfs Team,
>

That's me.

> Can you please review/include the attached patch into squashfs-tools ?
> The mentioned changes are needed by AppImageKit.
>

Well from a quick look at the attached patches, I see no reason
not to pull them.

I'll have a look over at github because is the first I've seen of these
patches, and I'll see if I can pull them from there.

If not, I'll ask the author to send them to me at this address.

Phillip@lougher.demon.co.uk is dead because my ISP arbitrarily ceased
email support sometime ago (and I'd had that email for 20 years).
It would appear phillip@squashfs,org.uk is still redirected to that,
and I'll fix that.

Phillip

> What is AppImage
>
> The AppImage format is a format for packaging applications in a way that allows
> them to run on a variety of different target systems (base operating systems,
> distributions) without further modification.
>
> https://en.wikipedia.org/wiki/AppImage
>
> AppImageKit is a concrete implementation of the AppImage format and provides
> tools such as appimagetool and appimaged for conveniently handling AppImages.
>
> https://github.com/probonopd/AppImageKit
>
>
>
> --
> Ritesh Raj Sarraf
> RESEARCHUT - http://www.researchut.com
> "Necessity is the mother of invention."

**Comment by rickysarraf** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Hello Phillip,

Good to hear from you. When the emails bounced, I was wondering if you are still
active or not.

On Tue, 2017-01-31 at 02:57 +0000, Phillip Lougher wrote:
> 
> The repo at kernel.org is the official repo.  It hasn't been updated in a
> while
> simply because the tools are considered stable, and I have not had a burning
> desire to add new features for a while.
> 

Cool. Thanks.

> > 
> > Can you please review/include the attached patch into squashfs-tools ?
> > The mentioned changes are needed by AppImageKit.
> > 
> 
> Well from a quick look at the attached patches, I see no reason
> not to pull them.
> 
> I'll have a look over at github because is the first I've seen of these
> patches, and I'll see if I can pull them from there.
> 
> If not, I'll ask the author to send them to me at this address.
> 

Great. The author (probono AT puredarwin.org) is copied on this email.

> Phillip@lougher.demon.co.uk is dead because my ISP arbitrarily ceased
> email support sometime ago (and I'd had that email for 20 years).
> It would appear phillip@squashfs,org.uk is still redirected to that,
> and I'll fix that.
> 

I used to maintain my own mail server but eventually gave up (amount of spam
kept the machine alive all the while) and switched to GApps.


- -- 
Ritesh Raj Sarraf
RESEARCHUT - http://www.researchut.com
"Necessity is the mother of invention."
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCgAdFiEEQCVDstmIVAB/Yn02pjpYo/LhdWkFAliQYmYACgkQpjpYo/Lh
dWlblg/8DxMhZEDN8CnL1ZslyIdlTtz9aFPQ73qjNSHvNKx+zO0KhgOC7dSbB/uU
BPlNBXlGcJ9q4rIVcgxEmETIUpag4aZND8bkfnX4rVffq8/Q0rPHNjknlkT8Wvah
UjsyqIBP5PgAlyJKNr2ukl5TXGk+OTpcAnWgZBEdsrJbFBF/GxRnGkHxr7DZEjkK
D2gPs2GTK6SF21+Mj2nkuNWzf2ofX1amf5ZQyOoTOOO0WBRquTQIKDGk5eM13+5o
+yP16UGUD60ADJfeSYXSAbLCRzIRNPc++9jmcyccpTdU5zSYGzeO5+UyH7wlTF5Q
lHPA0W5ss7xq2bW2iDM/+iPB75YY9iHGXsyWt66sHZ28BFR32KdoxzmQlKnV2+cT
QOBSDbGto+f64CbN+ufLROIu5g89G6LNOu1PIH6KVY4f7z2zCZK8VRvcL/XQschp
8+jc67fSv+go8yOY2C3VG1JYSds7vsb7TyJZQKxbEMQC0vkoO5R6olbCMNb+8O1A
kPk0eozr/GnWm1/xqXUMT9ZEhqeicQeTa2MXnPWypgfJnTxbkA4PuDUid1YUtpJ0
xNFJOSKu6SdFOgH2GkO8nXFlB8DOfvlulATdUJJyHa3ThPzo3EGPss9vZU8Gj34g
CJBovBqkNp3Azu26ksQrEMK2UP/BF9LaiKpPgTrhLkULWnH7nBQ=
=RCdF
-----END PGP SIGNATURE-----

**Comment by probonopd** [comment]
Is there any action required from my side at this time?

**Comment by probonopd** [comment]
Yay, it just was merged: https://github.com/plougher/squashfs-tools/pull/13#event-956250674

**Comment by rickysarraf** [comment]
Great news. I'll soon look at the debian/ packaging again.

**Comment by rickysarraf** [comment]
http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=854901
http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=854902

For anyone interested to help/collaborate, please feel free to drop me a note here, or on the referenced Debian ITPs.

**Comment by probonopd** [comment]
Thank you @rickysarraf. Minor disagreement, I don't think "Flatpak and Snaps solve similar problems". They don't allow users to download a single file from an application's download page, and run that on most Linux distributions without changes to the system.

**Comment by probonopd** [comment]
ping @ScarlettGatelyClark on http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=854902

**Comment by rickysarraf** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

On Sat, 2017-02-11 at 10:07 -0800, probonopd wrote:
> Thank you @rickysarraf. Minor disagreement, I don't think "Flatpak and Snaps
> solve similar problems". They don't allow users to download a single file from
> an application's download page, and run that on most Linux distributions
> without changes to the system.

THanks. I will pick your corrections when I start working on it.

And one reason why I like AppImage over Flatpak/Snaps/etc is really its
simplicity to "Just Run". Once I'm done with the package's inclusion into
Debian, I'd like to explore possibilities with Firejail/cgroup integration.

- -- 
Ritesh Raj Sarraf
RESEARCHUT - http://www.researchut.com
"Necessity is the mother of invention."
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCgAdFiEEQCVDstmIVAB/Yn02pjpYo/LhdWkFAlifVloACgkQpjpYo/Lh
dWmaYQ//RebFMejQgZfyv3RWdX/8P/6GjxlQkyBIR5l9eF++yhHogSveSt9xT7VW
si+meR5zOkUNXqAOiI0vrnnL+wwWMLdtamBOaAlWCVqQQwGn3kdlY+0TsHj+ilMd
6f0MOb3mayTL22Bb7p0yRgp8P7PZoFJBcpYShXr7oCK2hGLy4m0JbdXt592/Ce2Y
hLmpsrfPVlSXECVsw6m+jsB+4vY8/LkcFMAxDMczac5SC1jtAoMzqqUPILBbkkuA
oKdjjSFYQCZYzQ4zxvVJIQTMAgzLgDZLkDduyEJ9vXtMG/bY9g7Ggn7gniVf2Dm5
04KQMWpP5Wmdb0ZlltHf3Umpv1dYmEjehWcCq1/+f5tKzq2Z+kN5WlFEJvcwx3Zx
V/p2OZ4L2184znR6L5LZZfDtcZjrLJXqdJ9FcOFbevLE3lm0O1+H7WCAUqQimRj/
u4cK5S9+nR+8TyOAkyYDqpxeL1LIar9211H90ak3nfiJLoTcETZSAPF3Mo81+xrk
pJUDfSo+iZVL7wx5bkJvlUGFuAwVC+0+ffItC5BZNitf9mRJnAwN2X7lSaC6pKKJ
XNaNTrk2Ls0TAMAHAMsWeEgChYIYW6x34hAnw3lihtla43zupMQL95iHz/Ijejg4
dy/+5BnMAI+LfQDobSVG4f7yZL+knKIMbBed/6ARPUMPTU9cf5U=
=5tgh
-----END PGP SIGNATURE-----

**Comment by probonopd** [comment]
Also see https://build.opensuse.org/package/show?project=devel%3Atools%3Abuilding&package=appimagekit

**Comment by rickysarraf** [comment]
@probonopd I went off for quite some time. Real life priorities came in.

I have packaged the squashfuse dependency as a package now. But now, with version 9, I see the entire Cmake based build to have vanished. Is build.sh the way upstream way to build appimagekit ?

**Comment by TheAssassin** [comment]
Yes, build.sh is the most upstream way to create binaries. Though, it is pretty likely that you will need only parts of it, so you should think about copy-pasting to your own script for building only the parts you need.

I am working to create and improve CMake configs for the AppImage ecosystem (primarily for tool development and not necessarily for release buils), suggestions welcome.

**Comment by rickysarraf** [comment]
@TheAssassin You guys should have revived the build.sh script such that the actual compilation and 3rd party library fetches were separate.

@probonopd You still applying squashfuse patches. Weren't these included upstream the last time we visited this topic ?

```
cd squashfuse
if [ ! -e ./ll.c.orig ]; then
  patch -p1 --backup < ../squashfuse.patch
  patch -p1 --backup < ../squashfuse_dlopen.patch
fi
if [ ! -e ./squashfuse_dlopen.c ]; then
  cp ../squashfuse_dlopen.c .
fi
if [ ! -e ./squashfuse_dlopen.h ]; then
  cp ../squashfuse_dlopen.h .
fi
```

@probonopd Also, that opensuse build link you've shared, is broken.

**Comment by rickysarraf** [comment]
So this is the complete build rules for appimagekit, excluding the dependencies. If I am to package it for Debian, I'd just adapt these. But it'd be nicer if this was done upstream.

```
# Compile runtime but do not link

$CC -DVERSION_NUMBER=\"$(git describe --tags --always --abbrev=7)\" -I../squashfuse/ -D_FILE_OFFSET_BITS=64 -g -Os -c ../runtime.c

# Prepare 1024 bytes of space for updateinformation
printf '\0%.0s' {0..1023} > 1024_blank_bytes

objcopy --add-section .upd_info=1024_blank_bytes \
        --set-section-flags .upd_info=noload,readonly runtime.o runtime2.o

objcopy --add-section .sha256_sig=1024_blank_bytes \
        --set-section-flags .sha256_sig=noload,readonly runtime2.o runtime3.o

# Now statically link against libsquashfuse_ll, libsquashfuse and liblzma
# and embed .upd_info and .sha256_sig sections
$CC -o runtime ../elf.c ../notify.c ../getsection.c runtime3.o \
    ../squashfuse/.libs/libsquashfuse_ll.a ../squashfuse/.libs/libsquashfuse.a ../squashfuse/.libs/libfuseprivate.a \
    -L../xz-5.2.3/build/lib -Wl,-Bdynamic -lpthread -lz -Wl,-Bstatic -llzma -Wl,-Bdynamic -ldl
$STRIP runtime

# Test if we can read it back
readelf -x .upd_info runtime # hexdump
readelf -p .upd_info runtime || true # string

# The raw updateinformation data can be read out manually like this:
HEXOFFSET=$(objdump -h runtime | grep .upd_info | awk '{print $6}')
HEXLENGTH=$(objdump -h runtime | grep .upd_info | awk '{print $3}')
dd bs=1 if=runtime skip=$(($(echo 0x$HEXOFFSET)+0)) count=$(($(echo 0x$HEXLENGTH)+0)) | xxd

# Insert AppImage magic bytes

printf '\x41\x49\x02' | dd of=runtime bs=1 seek=8 count=3 conv=notrunc

# Convert runtime into a data object that can be embedded into appimagetool

ld -r -b binary -o data.o runtime

# Test if we can read it back
readelf -x .upd_info runtime # hexdump
readelf -p .upd_info runtime || true # string

# The raw updateinformation data can be read out manually like this:
HEXOFFSET=$(objdump -h runtime | grep .upd_info | awk '{print $6}')
HEXLENGTH=$(objdump -h runtime | grep .upd_info | awk '{print $3}')
dd bs=1 if=runtime skip=$(($(echo 0x$HEXOFFSET)+0)) count=$(($(echo 0x$HEXLENGTH)+0)) | xxd

# Convert runtime into a data object that can be embedded into appimagetool

ld -r -b binary -o data.o runtime

# Compile appimagetool but do not link - glib version

$CC -DVERSION_NUMBER=\"$(git describe --tags --always --abbrev=7)\" -D_FILE_OFFSET_BITS=64 -I../squashfuse/ \
    $(pkg-config --cflags glib-2.0) -g -Os ../getsection.c  -c ../appimagetool.c

# Now statically link against libsquashfuse - glib version
if [ $STATIC_BUILD -eq 1 ]; then
  # statically link against liblzma
  $CC -o appimagetool data.o appimagetool.o ../elf.c ../getsection.c -DENABLE_BINRELOC ../binreloc.c \
    ../squashfuse/.libs/libsquashfuse.a ../squashfuse/.libs/libfuseprivate.a \
    -L../xz-5.2.3/build/lib \
    -Wl,-Bdynamic -ldl -lpthread \
    -Wl,--as-needed $(pkg-config --cflags --libs glib-2.0) -lz -Wl,-Bstatic -llzma -Wl,-Bdynamic
else
  # dinamically link against distro provided liblzma
  $CC -o appimagetool data.o appimagetool.o ../elf.c ../getsection.c -DENABLE_BINRELOC ../binreloc.c \
    ../squashfuse/.libs/libsquashfuse.a ../squashfuse/.libs/libfuseprivate.a \
    -Wl,-Bdynamic -ldl -lpthread \
    -Wl,--as-needed $(pkg-config --cflags --libs glib-2.0) -lz -llzma
fi

# Version without glib
# cc -D_FILE_OFFSET_BITS=64 -I ../squashfuse -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -g -Os -c ../appimagetoolnoglib.c
# cc data.o appimagetoolnoglib.o -DENABLE_BINRELOC ../binreloc.c ../squashfuse/.libs/libsquashfuse.a ../squashfuse/.libs/libfuseprivate.a -Wl,-Bdynamic -ldl -lpthread -lz -Wl,-Bstatic -llzma -Wl,-Bdynamic -o appimagetoolnoglib

# Compile and link digest tool

if [ $STATIC_BUILD -eq 1 ]; then
  $CC -o digest ../getsection.c ../digest.c -I../openssl-1.1.0c/build/include -L../openssl-1.1.0c/build/lib \
    -Wl,-Bstatic -lssl -lcrypto -Wl,-Bdynamic -lz -ldl
else
  $CC -o digest ../getsection.c ../digest.c -Wl,-Bdynamic -lssl -lcrypto -lz -ldl
fi

$STRIP digest

# Compile and link validate tool

if [ $STATIC_BUILD -eq 1 ]; then
  $CC -o validate ../getsection.c ../validate.c -I../openssl-1.1.0c/build/include -L../openssl-1.1.0c/build/lib \
    -Wl,-Bstatic -lssl -lcrypto -Wl,-Bdynamic -Wl,--as-needed $(pkg-config --cflags --libs glib-2.0) -lz -ldl
else
  $CC -o validate ../getsection.c ../validate.c -Wl,-Bdynamic -lssl -lcrypto \
    -Wl,--as-needed $(pkg-config --cflags --libs glib-2.0) -lz -ldl
fi

$STRIP validate

# AppRun
$CC ../AppRun.c -o AppRun

# check for libarchive name
have_libarchive3=0
archive_n=
if printf "#include <archive3.h>\nint main(){return 0;}" | cc -w -O0 -xc - -Wl,--no-as-needed -larchive3 2>/dev/null ; then
  have_libarchive3=1
  archive_n=3
fi
rm -f a.out

# appimaged, an optional component
if [ $STATIC_BUILD -eq 1 ]; then
  $CC -std=gnu99 -o appimaged -I../squashfuse/ ../getsection.c ../notify.c ../elf.c ../appimaged.c \
    -D_FILE_OFFSET_BITS=64 -DHAVE_LIBARCHIVE3=$have_libarchive3 -DVERSION_NUMBER=\"$(git describe --tags --always --abbrev=7)\" \
    ../squashfuse/.libs/libsquashfuse.a ../squashfuse/.libs/libfuseprivate.a \
    -L../xz-5.2.3/build/lib -I../inotify-tools-3.14/build/include -L../inotify-tools-3.14/build/lib \
    -Wl,-Bstatic -linotifytools -Wl,-Bdynamic -larchive${archive_n} \
    -Wl,--as-needed \
    $(pkg-config --cflags --libs glib-2.0) \
    $(pkg-config --cflags --libs gio-2.0) \
    $(pkg-config --cflags --libs cairo) \
    -ldl -lpthread -lz -Wl,-Bstatic -llzma -Wl,-Bdynamic
else
  $CC -std=gnu99 -o appimaged -I../squashfuse/ ../getsection.c ../notify.c ../elf.c ../appimaged.c \
    -D_FILE_OFFSET_BITS=64 -DHAVE_LIBARCHIVE3=$have_libarchive3 -DVERSION_NUMBER=\"$(git describe --tags --always --abbrev=7)\" \
    ../squashfuse/.libs/libsquashfuse.a ../squashfuse/.libs/libfuseprivate.a \
    -Wl,-Bdynamic -linotifytools -larchive${archive_n} \
    -Wl,--as-needed \
    $(pkg-config --cflags --libs glib-2.0) \
    $(pkg-config --cflags --libs gio-2.0) \
    $(pkg-config --cflags --libs cairo) \
    -ldl -lpthread -lz -llzma
fi

cd ..

# Strip and check size and dependencies

rm build/*.o build/1024_blank_bytes
$STRIP build/* 2>/dev/null
chmod a+x build/*
ls -lh build/*
for FILE in $(ls build/*) ; do
  echo "$FILE"
  ldd "$FILE" || true
done

bash -ex "$HERE/build-appdirs.sh"

ls -lh

mkdir -p out
cp -r build/* ./*.AppDir out/
```

**Comment by probonopd** [comment]
> @@TheAssassin You guys should have revived the build.sh script such that the actual compilation and 3rd party library fetches were separate.

I 
Wouldn't it be sufficient to (pseudo-code)

```
if not (building_debian_package){
  do stuff that you don't need
}
```

I'd happily merge a PR with such `if` statements as long as it doesn't alter my builds and only gets in effect if your debian source package is used.

> @probonopd You still applying squashfuse patches. Weren't these included upstream the last time we visited this topic ?

Were all of them? I was only aware of the offset patch having been accepted.
In any case, what we have now is working for me. PR welcome if you think it needs to be changed.

> @probonopd Also, that opensuse build link you've shared, is broken.

https://build.opensuse.org/package/show/devel:tools:building/appimagekit seems to be the official one now.

**Comment by probonopd** [comment]
> So this is the complete build rules for appimagekit

Isn't this the build.sh script with some extras removed? If so, please use `if` statements as described above.

**Comment by TheAssassin** [comment]
As you might have noticed, there's also a CMake configuration that we could extend and modify to work for you, @rickysarraf. If you just need appimaged, let me add the configuration and we're good to goo.

**Comment by azubieta** [comment]
Please move to the appimaged repo.
